packages/react/src/analytics/integrations/Castle/CastleV2.js

```diff
@@ -0,0 +1,328 @@
+
+    this.validateOptions(options);
+    this.applyOptions(options);
+  }
+
+  /**
+   * Method that will validate each important option and throw an error if necessary.
+   *
+   * @param {object} options - Integration options.
+   */
+  validateOptions(options) {
+    if (!options.pk) {
+      throw new Error(
+        `${CASTLE_MESSAGE_PREFIX} Failed to initialize Castle. Please make sure a valid 'pk: string (Publishable Key)' is passed via the integration options.`,
+      );
+    }
+
+    if (
+      options.configureHttpClient &&
+      typeof options.configureHttpClient !== 'function'
+    ) {
+      throw new TypeError(
+        `${CASTLE_MESSAGE_PREFIX} Invalid option 'configureHttpClient'. Please make sure a valid function to configure the necessary headers.`,
+      );
+    }
+
+    if (
+      this.clientIdHeaderName &&
+      typeof this.clientIdHeaderName !== 'string'
+    ) {
+      throw new TypeError(
+        `${CASTLE_MESSAGE_PREFIX} Invalid option 'clientIdHeaderName'. Please make sure a valid string for the castle header name.`,
+      );
+    }
+  }
+
+  /**
+   * After the validations, this method will apply the options passed in and install the interceptors.
+   *
+   * @param {object} options - Integration options.
+   */
+  applyOptions(options) {
+    try {
+      this.castleJS.configure(options);
+
+      this.installInterceptors(options);
+    } catch (error) {
+      throw new Error(
+        `${CASTLE_MESSAGE_PREFIX} Failed to initialize Castle. Error: ${error}`,
+      );
+    }
+  }
+
+  /**
+   * Installs the necessary request interceptors on the axios client.
+   *
+   * @param {object} options - Integration options.
+   *
+   * @async
+   */
+  async installInterceptors(options) {
+    if (options.configureHttpClient) {
+      await options.configureHttpClient(this.castleJS.createRequestToken);
+
+      this.isInterceptorAttached = true;
+
+      return;
+    }
+
+    // install the default interceptors on our core client
+    this.httpClientInterceptor = this.httpClient.interceptors.request.use(
+      this.onBeforeRequestFullfil,
+      this.onBeforeRequestReject,
+    );
+
+    this.isInterceptorAttached = true;
+  }
+
+  /**
+   * Callback that is used on the Axios interceptor to add the correct Castle token header.
+   *
+   * @async
+   *
+   * @param {object} config - Axios config object.
+   *
+   * @returns {object} - The modified Axios config object.
+   */
+  onBeforeRequestFullfil = async config => {
+    const requestTokenValue = await this.castleJS.createRequestToken();
+
+    config.headers[this.clientIdHeaderName] = requestTokenValue;
+
+    return config;
+  };
+
+  /**
+   * Callback that is used on the Axios interceptor to handle any error when setting the Castle header.
+   *
+   * @async
+   *
+   * @param {object|string} error - The error itself.
+   *
+   * @returns {Promise} - The rejected promise.
+   */
+  onBeforeRequestReject = async error => {
```

Comment:
This method can be removed as by default axios will reject the promise automatically if there was an error with the request.
To do so, you can change the attach interceptors instruction to this:

```js
// install the default interceptors on our core client
    this.httpClientInterceptor = this.httpClient.interceptors.request.use(
      this.onBeforeRequestFullfil,
      null (or undefined),
    );

```